apps/test/unit/weblab/brambleUtilsTest.js

```diff
@@ -0,0 +1,128 @@
+import {expect} from '../../util/reconfiguredChai';
+import sinon from 'sinon';
+import {removeDisallowedHtmlContent} from '@cdo/apps/weblab/brambleUtils';
+
+// This regex is meant to mimic the way we disallow certain HTML tags in WebLab.js.
+const REGEX = new RegExp(`<(script|a)[\\s\\S]*\\/(script|a)*>`, 'gi');
+
+describe('removeDisallowedHtmlContent', () => {
+  let fileSystemStub, brambleProxyStub, callbackSpy;
+
+  beforeEach(() => {
+    fileSystemStub = {
+      readFile: sinon.spy(),
+      writeFile: sinon.spy()
+    };
+    brambleProxyStub = {
+      enableReadonly: sinon.spy()
+    };
+    callbackSpy = sinon.spy();
+  });
+
+  it('no-ops if path does not point to an HTML file', () => {
+    removeDisallowedHtmlContent(
+      fileSystemStub,
+      brambleProxyStub,
+      '/style.css',
+      REGEX,
+      callbackSpy
+    );
+
+    expect(fileSystemStub.readFile).to.not.have.been.called;
+    expect(fileSystemStub.writeFile).to.not.have.been.called;
+    expect(brambleProxyStub.enableReadonly).to.not.have.been.called;
+    expect(callbackSpy).to.have.been.calledOnceWith('/style.css');
+  });
+
+  it('no-ops if reading file errored', () => {
+    fileSystemStub.readFile = sinon
+      .stub()
+      .callsFake((path, encoding, callback) => {
+        callback(new Error('uh oh!'), null);
+      });
+
+    removeDisallowedHtmlContent(
+      fileSystemStub,
+      brambleProxyStub,
+      '/index.html',
+      REGEX,
+      callbackSpy
+    );
+
+    expect(fileSystemStub.readFile).to.have.been.calledOnce;
+    expect(fileSystemStub.writeFile).to.not.have.been.called;
+    expect(brambleProxyStub.enableReadonly).to.not.have.been.called;
+    expect(callbackSpy).to.have.been.calledOnceWith('/index.html');
+  });
+
+  it('no-ops if file contents do not contain disallowed content', () => {
+    fileSystemStub.readFile = sinon
+      .stub()
+      .callsFake((path, encoding, callback) => {
+        const fileData = `<!DOCTYPE html>
+        <html>
+          <body>
+            <p>Important paragraph.</p>
+          </body>
+        </html>
+        `;
+        callback(null, fileData);
+      });
+
+    removeDisallowedHtmlContent(
+      fileSystemStub,
+      brambleProxyStub,
+      '/index.html',
+      REGEX,
+      callbackSpy
+    );
+
+    expect(fileSystemStub.readFile).to.have.been.calledOnce;
+    expect(fileSystemStub.writeFile).to.not.have.been.called;
+    expect(brambleProxyStub.enableReadonly).to.not.have.been.called;
+    expect(callbackSpy).to.have.been.calledOnceWith('/index.html');
+  });
+
+  it('writes the file without disallowed content', () => {
+    // Based on REGEX, both <script> and <a> tags should be removed from the template below.
+    const invalidFileData = `<!DOCTYPE html>
+    <html>
+      <body>
+        <script src="index.js">
+        </script>
+        <a href="/some.url">I will be deleted!</a>
+        <div>divs are allowed</div>
+      </body>
+    </html>
+    `;
```

Comment:
Just a note that I believe the regex being used here will remove all content on a matching line. If possible we might want to adjust the rule and add a test for this use case

```
    const invalidFileData = `<!DOCTYPE html>
    <html>
      <body>
        <script src="index.js">
        </script>
        I should not be deleted <a href="/some.url">I will be deleted!</a>
        <div>divs are allowed</div>
      </body>
    </html>
    `;
```

Reply:
that case does work already, but definitely needs a unit test.

i'm going to co-opt this comment to discuss using DOMParser in place of a regex here, but let me know if this is easier to talk about in-person...

using DOMParser here works great except for preserving whitespace, as you thought. there are a couple of options that i think we have here if we go this route:

```javascript
const domParser = new DOMParser();
const dom = domParser.parseFromString(data, "text/html");
const nodes = dom.querySelectorAll('script'); // TODO: query for all disallowed element types

for (var i = 0; i < nodes.length; i++) {
  nodes[i].parentElement.removeChild(nodes[i]);
}

// OPTION 1: use XMLSerializer
const serializer = new XMLSerializer();
data = serializer.serializeToString(dom);

// OPTION 2: manually insert content back into a string (caveat: the whitespace/newlines need some tweaking here)
data = `<!DOCTYPE html>\n<html>\n  <head>\n    ${dom.head.innerHTML}\n  </head>\n  <body>\n    ${dom.body.innerHTML}\n  </body>\n</html>`;
```

Option 1 would be ideal/less error-prone, but XMLSerializer makes some weird assumptions about how you want the resulting string to be formatted:
```html
<!-- XMLSerializer structure -->
<!DOCTYPE html><html xmlns="http://www.w3.org/1999/xhtml"><head>
    
  </head>
  <body>
    
  
</body></html>

<!-- Our structure (we control this, not Bramble) -->
<!DOCTYPE html>
<html>
  <head>
    
  </head>
  <body>
   
    
  </body>
</html>
```
...but we _could_ update our default HTML content to match this structure [here](https://github.com/code-dot-org/code-dot-org/blob/dcb822e1aeb495fe4286f7dbf8b8cf3ec888882b/apps/src/weblab/brambleHost.js#L419). that may be a bigger change than we want to make and may have cascading problems with indentation.

Option 2 is more error-prone, but we have more granular control over whitespace in the resulting string and would match our default HTML content structure.

i'm leaning toward option 2, but i'm interested to hear your thoughts or any alternatives i haven't considered

Reply:
@madelynkasula, For option 2, is there any concern with users not having `head` or `body` tags at all in their markup? I suppose in this case we could just add those for them?

In general I like this option. I wonder if in the case of our filter not finding any disallowed tags, we could skip the routine of stringifying the parsed DOM and simply return the user content as they submitted it. That way we would only risk whitespace formatting issues on the (probably pretty rare) times that the user has added a `script` or `iframe`

Reply:
> is there any concern with users not having head or body tags at all in their markup? I suppose in this case we could just add those for them?

that's a good question. i don't think this is a concern (or, at least, wouldn't be surprising/alarming for users) because every time they create an HTML file, we add a head and body for them. i think many users assume that the code we've generated for them is required in some way, so i don't imagine users removing head/body tags often. however, if we get feedback from users about this, we could detect whether the original file data had a head and body and only include head/body if it was previously there.

also, agreed that if we don't find any disallowed tags we should exit early -- i have that built in for the regex approach already so that we aren't constantly writing to the file unless we really need to, and it'll be easy to add for this approach too.

i'll move forward with option 2 and ping *labs for another review once it's ready!